### PR TITLE
Do not add objects to the DataTree

### LINKED ISSUE #2265: TypeError: Cannot use 'in' operator to search for 'ENTITY_TYPE' in
Sentry Issue: [APPSMITH-EN](https://sentry.io/organizations/appsmith/issues/2098830073/?referrer=github_integration)

```
TypeError: Cannot use 'in' operator to search for 'ENTITY_TYPE' in 
  at None (utils/autocomplete/dataTreeTypeDefCreator.ts:19:9)
  at Array.forEach (<anonymous>)
  at dataTreeTypeDefCreator (utils/autocomplete/dataTreeTypeDefCreator.ts:17:25)
  at new C (utils/autocomplete/TernServer.ts:61:25)
  at helper (components/editorComponents/CodeEditor/hintHelpers.ts:11:22)
...
(18 additional frame(s) were not displayed)
```

### PR DESCRIPTION
**Pull request by ofpiyush**
## Description

Fixes #2265
## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

N/A

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/client/src/sagas/evaluationsSaga.ts [hunk 1]
```diff
@@ -117,10 +117,11 @@ export function* evaluateSingleValue(
 ) {
   if (evaluationWorker) {
     const dataTree = yield select(getDataTree);
-    dataTree[EXECUTION_PARAM_KEY] = executionParams;
     evaluationWorker.postMessage({
       action: EVAL_WORKER_ACTIONS.EVAL_SINGLE,
-      dataTree,
+      dataTree: Object.assign({}, dataTree, {
+        [EXECUTION_PARAM_KEY]: executionParams,
+      }),
```
Comment (ofpiyush) [inline app/client/src/sagas/evaluationsSaga.ts:124]: Actual fix

File: app/client/src/reducers/evaluationReducers/treeReducer.ts [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
+import { createImmerReducer } from "utils/AppsmithUtils";
 import { DataTree } from "entities/DataTree/dataTreeFactory";
 import { ReduxAction, ReduxActionTypes } from "constants/ReduxActionConstants";
 
 export type EvaluatedTreeState = DataTree;
 
 const initialState: EvaluatedTreeState = {};
 
-const evaluatedTreeReducer = createReducer(initialState, {
+const evaluatedTreeReducer = createImmerReducer(initialState, {
```
Comment (ofpiyush) [inline app/client/src/reducers/evaluationReducers/treeReducer.ts:9]: This line ensures that this never happens again.

File: app/client/src/reducers/index.tsx [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import { combineReducers } from "redux";
 import entityReducer from "./entityReducers";
 import uiReducer from "./uiReducers";
-import evaluationsReducer from "./evalutationReducers";
+import evaluationsReducer from "./evaluationReducers";
```
Comment (ofpiyush) [inline app/client/src/reducers/index.tsx:4]: random typo fix 😄

File: app/client/src/utils/autocomplete/dataTreeTypeDefCreator.ts [hunk 4]
```diff
@@ -16,7 +16,7 @@ export const dataTreeTypeDefCreator = (dataTree: DataTree) => {
   };
   Object.keys(dataTree).forEach(entityName => {
     const entity = dataTree[entityName];
-    if ("ENTITY_TYPE" in entity) {
+    if (entity && "ENTITY_TYPE" in entity) {
```
Comment (ofpiyush) [inline app/client/src/utils/autocomplete/dataTreeTypeDefCreator.ts:19]: Last ditch sanity fix to make the check same everywhere

File: app/client/cypress/integration/Smoke_TestSuite/Binding/Bind_TableTextPagination_spec.js [hunk 5]
```diff
@@ -23,7 +23,7 @@ describe("Test Create Api and Bind to Table widget", function() {
     /**Bind Table with Textwidget with selected row */
     cy.SearchEntityandOpen("Text1");
     cy.testJsontext("text", "{{Table1.selectedRow.url}}");
-    cy.get(commonlocators.editPropCrossButton).click();
+    cy.SearchEntityandOpen("Table1");
```
Comment (ofpiyush) [inline app/client/cypress/integration/Smoke_TestSuite/Binding/Bind_TableTextPagination_spec.js:26]: Change test back to catch this kind of bug if it ever creeps up again.

File: app/client/src/utils/autocomplete/dataTreeTypeDefCreator.ts (hunk 4 above)
Comment (mohanarpit) [reply app/client/src/utils/autocomplete/dataTreeTypeDefCreator.ts:19]: Given that we are doing `Object.assign` in `evaluationSagas`, this null condition check will always be true right? #JustConfirming.

Comment (ofpiyush) [reply app/client/src/utils/autocomplete/dataTreeTypeDefCreator.ts:19]: Everywhere else in the codebase we did a `entity && "ENTITY_TYPE" in entity` check, so I did this more for consistency than anything else.

**Review by mohanarpit — APPROVED** [review]

**Comment by ofpiyush** [comment]
Explanation of the bug:

- Datatree was getting an extra element which evaluated to an empty string for a very short time.
- If hinting helper got triggered in this time, it would loop over the `dataTree`.
- If this loop happened within the split-second, it would cause a crash.

**Review by rishabhsaxena — APPROVED** [review]

**Pull request merged**